### DISCUSSION TITLE
How to use the Arena

### DISCUSSION THREAD

**Discussion by Lin-Mao** [AMReX-Codes/amrex#2470] [General]
I am new to AMReX and I am going to write some test cases of Arena. So what should I do to get started with this API? Do any guys have a available examples or posts about it? Thanks so much for your help!!!

**Comment by WeiqunZhang** [comment]
What aspects of `Arena`  do you want to test?  `Arena`'s interface is mainly just
```
virtual void* alloc (std::size_t sz);
virtual void free (void* pt)
```

**Reply by Lin-Mao** [reply]
Thanks so much for your reply, Weiqun! I am going to write a test case to capture and analyze the GPU sub-allocation(It's not a real allocation, but just registering a memory range from an allocated GPU memory). I am not sure if `Arena::alloc` can do this job. I think `Arena::alloc` is an actual allocation operation. If that, could you please tell me which api can do that job in AMReX? I would appreciate it.

**Reply by WeiqunZhang** [reply]
`Arena` is an abstract class.  If you use `CArena` and call `alloc`, it will eventually call `cudaMallocManaged` or `cudaMalloc` or something else depending on the setting.  Does that answer your question?
```
AreaInfo ai{}; // By default, it uses managed memory
// ai.SetDeviceMemory(); // Use device memory instead of managed memory
CArena my_arena(0, ai);
my_arena.alloc(size);
```

**Reply by Lin-Mao** [reply]
Thank you! I think you've feed my question. By the way, could you please also tell me which code example in amrex-tutorials using `CArena::alloc` in that way, so that I can quickly get started? Thank you again.

**Reply by WeiqunZhang** [reply]
AMReX users typically do not call `alloc` directly.  So I don't think we have any example for it.  But it's just
```
amrex::Initialize(argc,argv);
{
    AreaInfo ai{}; // By default, it uses managed memory
    // ai.SetDeviceMemory(); // Use device memory instead of managed memory
    CArena my_arena(0, ai);
    my_arena.alloc(size);
}
amrex::Finalize();
```

**Reply by Lin-Mao** [reply]
> AMReX users typically do not call alloc directly.

What did you actually mean by saying this? Is that to say there are high level APIs to encapsulate the `CArena::alloc`, and users just need to call high level APIs? 
In addition, thank you for your demo code, it helps me a lot!

**Reply by WeiqunZhang** [reply]
Right, it's like C++ users are encouraged to use `std::vector`, instead of `new`.

**Reply by Lin-Mao** [reply]
Got it. In this way, could you please tell me one or more APIs encapsulate `CArena::alloc`, and one or more code examples in amrex-tutorials  for the corresponding APIs? I would appreciate it.

**Reply by WeiqunZhang** [reply]
The constructor of `MultiFab` will eventually call `alloc`.  Almost every tutorial uses it.

**Reply by Lin-Mao** [reply]
Got it. Thanks so much for your help. I think I have some ideas of what to do next.
